lib/types/jtd-schema.ts

```diff
@@ -147,3 +147,69 @@ export type JTDSchemaType<T, D extends Record<string, unknown> = Record<string,
   // TODO these should only be allowed at the top level
   definitions?: {[K in keyof D]: JTDSchemaType<D[K], D>}
 }
+
+type JTDDataDef<S, D extends Record<string, unknown>> =
+  | (// ref
+    S extends {ref: string}
+      ? JTDDataDef<D[S["ref"]], D>
+      : // type
+      S extends {type: NumberType}
+      ? number
+      : S extends {type: "string"}
+      ? string
+      : S extends {type: "timestamp"}
+      ? string | Date
+      : // enum
+      S extends {enum: readonly (infer E)[]}
+      ? string extends E
+        ? never
+        : [E] extends [string]
+        ? E
+        : never
+      : // elements
+      S extends {elements: infer E}
+      ? JTDDataDef<E, D>[]
+      : // properties
+      S extends {
+          properties: Record<string, unknown>
+          optionalProperties?: Record<string, unknown>
+          additionalProperties?: boolean
+        }
+      ? {-readonly [K in keyof S["properties"]]-?: JTDDataDef<S["properties"][K], D>} &
+          {
+            -readonly [K in keyof S["optionalProperties"]]+?: JTDDataDef<
+              S["optionalProperties"][K],
+              D
+            >
+          } &
+          ([S["additionalProperties"]] extends [true] ? Record<string, unknown> : unknown)
+      : S extends {
+          properties?: Record<string, unknown>
+          optionalProperties: Record<string, unknown>
+          additionalProperties?: boolean
+        }
+      ? {-readonly [K in keyof S["properties"]]-?: JTDDataDef<S["properties"][K], D>} &
+          {
+            -readonly [K in keyof S["optionalProperties"]]+?: JTDDataDef<
+              S["optionalProperties"][K],
+              D
+            >
+          } &
+          ([S["additionalProperties"]] extends [true] ? Record<string, unknown> : unknown)
```

Comment:
Possibly these two blocks can be refactored in some way... I will merge regardless.

Reply:
I can't think of anything major. Technically I could remove properties, since it's they're specified the first will match. However this was more in line with the spec which is why I kept it.

Do you have a preference?

Reply:
I was just wondering if the check than one of them is present could refactored, without changing anything. Let’s keep as is